src/java.base/share/classes/java/lang/reflect/AccessFlag.java

```diff
@@ -618,37 +575,227 @@ public enum Location {
-                                Set.of(PUBLIC, PRIVATE, PROTECTED,
-                                       STATIC, FINAL, INTERFACE, ABSTRACT,
-                                       SYNTHETIC, ANNOTATION, ENUM)),
-                          entry(Location.METHOD_PARAMETER,
-                                Set.of(FINAL, SYNTHETIC, MANDATED)),
-                          entry(Location.MODULE,
-                                Set.of(OPEN, SYNTHETIC, MANDATED)),
-                          entry(Location.MODULE_REQUIRES,
-                                Set.of(TRANSITIVE, STATIC_PHASE, SYNTHETIC, MANDATED)),
-                          entry(Location.MODULE_EXPORTS,
-                                Set.of(SYNTHETIC, MANDATED)),
-                          entry(Location.MODULE_OPENS,
-                                Set.of(SYNTHETIC, MANDATED)));
+    // Will take extra args in the future for valhalla switch
+    private static AccessFlag[] findDefinition(Location location) {
+        return switch (location) {
+            case CLASS -> CLASS_FLAGS;
+            case FIELD -> FIELD_FLAGS;
+            case METHOD -> METHOD_FLAGS;
+            case INNER_CLASS -> INNER_CLASS_FLAGS;
+            case METHOD_PARAMETER -> METHOD_PARAMETER_FLAGS;
+            case MODULE -> MODULE_FLAGS;
+            case MODULE_REQUIRES -> MODULE_REQUIRES_FLAGS;
+            case MODULE_EXPORTS -> MODULE_EXPORTS_FLAGS;
+            case MODULE_OPENS -> MODULE_OPENS_FLAGS;
+        };
+    }
+
+    private static final @Stable AccessFlag[] // Can use stable array and lazy init in the future
+            CLASS_FLAGS = createDefinition(PUBLIC, FINAL, SUPER, INTERFACE, ABSTRACT, SYNTHETIC, ANNOTATION, ENUM, MODULE),
+            FIELD_FLAGS = createDefinition(PUBLIC, PRIVATE, PROTECTED, STATIC, FINAL, VOLATILE, TRANSIENT, SYNTHETIC, ENUM),
+            METHOD_FLAGS = createDefinition(PUBLIC, PRIVATE, PROTECTED, STATIC, FINAL, SYNCHRONIZED, BRIDGE, VARARGS, NATIVE, ABSTRACT, STRICT, SYNTHETIC),
+            INNER_CLASS_FLAGS = createDefinition(PUBLIC, PRIVATE, PROTECTED, STATIC, FINAL, INTERFACE, ABSTRACT, SYNTHETIC, ANNOTATION, ENUM),
+            METHOD_PARAMETER_FLAGS = createDefinition(FINAL, SYNTHETIC, MANDATED),
+            MODULE_FLAGS = createDefinition(OPEN, SYNTHETIC, MANDATED),
+            MODULE_REQUIRES_FLAGS = createDefinition(TRANSITIVE, STATIC_PHASE, SYNTHETIC, MANDATED),
+            MODULE_EXPORTS_FLAGS = createDefinition(SYNTHETIC, MANDATED),
+            MODULE_OPENS_FLAGS = createDefinition(SYNTHETIC, MANDATED);
+
+    private static int undefinedMask(AccessFlag[] definition, int mask) {
+        assert definition.length == Character.SIZE;
+        int definedMask = 0;
+        for (int i = 0; i < Character.SIZE; i++) {
+            if (definition[i] != null) {
+                definedMask |= 1 << i;
+            }
+        }
+        return mask & ~definedMask;
+    }
+
+    private static final class AccessFlagSet extends AbstractSet<AccessFlag> {
```

Comment:
This should probably also define the spliterator method:
```java
@Override
public Spliterator<AccessFlag> spliterator() {
	return Spliterators.spliterator(
		this,
		Spliterator.DISTINCT	| Spliterator.ORDERED	| Spliterator.SIZED
		| Spliterator.NONNULL	| Spliterator.IMMUTABLE	| Spliterator.SUBSIZED
	);
}
```

Reply:
Hmm, I intentionally avoided implementing spliterator because there are too many opportunities to optimize - a fully optimized Spliterator implementaton should split an int mask. Can we do that in another patch? I need to check if MOAT tests spliterators too.

Reply:
Well, in this case, my suggestion isn’t about providing a fully optimised `Spliterator`, but about setting the correct characteristics.

Otherwise, its definition is exactly the same as in `Set::spliterator()`: https://github.com/openjdk/jdk/blob/ead6757486fb58f35f6e93b691519673bca1a549/src/java.base/share/classes/java/util/Set.java#L439-L442

Reply:
I think this probably gets its own RFE - Note that ImmutableCollections.SetN and Set12 (when there are 2 items) uses this default spliterator with no other flags set, even though it can set NONNULL and IMMUTABLE. Same for enum sets.